### PR TITLE
ENH: numba apply supports positional arguments passed as **kwargs

### PR DESCRIPTION
**Pull request by auderson**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

@lithomas1

This is a followup PR that enhances `prepare_function_arguments` introduced in #58767. `prepare_function_arguments` helps to move `**kwargs` to `*args` where possible using `inspect.Signature`.  And now it's also more reasonable to move kwargs check from `get_jit_arguments` to `prepare_function_arguments`.

ref: https://github.com/pandas-dev/pandas/pull/58767#issuecomment-2131310623
> We can also extend `prepare_function_arguments` by adding a option: `num_required_args`, so that all the numba functions' args & kwargs stuff can be handled by this function, instead of in `get_jit_arguments` . For example in groupby, we have `numba_func(group, group_index, *args)`, in this case `num_required_args = 2`:
> 
> https://github.com/pandas-dev/pandas/blob/3b48b17e52f3f3837b9ba8551c932f44633b5ff8/pandas/core/groupby/numba_.py#L102-L120

There are several places where the usage of `get_jit_arguments` is updated and `prepare_function_arguments` is added : `groupby.agg`, `groupby.transform`, `rolling.apply` and `df.apply`.

### CODE REVIEW

**Comment by auderson** [comment]
> This pull request is stale because it has been open for thirty days with no activity. Please [update](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#updating-your-pull-request) and respond to this comment if you're still interested in working on this.

Please do not mark as stale, this PR is ready for review.

**Comment by rhshadrach** [comment]
@auderson - can you merge main.

File: pandas/core/apply.py [hunk 1]
```diff
@@ -994,14 +994,15 @@ def wrapper(*args, **kwargs):
                 self.func,  # type: ignore[arg-type]
                 self.args,
                 self.kwargs,
+                1,
```
Comment (rhshadrach) [inline pandas/core/apply.py:997]: Can you pass this by name throughout; it is not clear what `1` is doing here.

File: pandas/core/util/numba_.py [hunk 2]
```diff
@@ -123,6 +109,8 @@ def prepare_function_arguments(
         user input positional arguments
     kwargs : dict
         user input keyword arguments
+    num_required_args : int
+        the number of required leading positional arguments for udf.
```
Comment (rhshadrach) [inline pandas/core/util/numba_.py:113]: I think this is the number of arguments that pandas (and not the user) will pass, right? So perhaps a better description is

> The number of positional arguments we will pass to udf. These are not supplied by the user.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
This is looking good to me. A few requests.

**Comment by rhshadrach** [comment]
@lithomas1 - friendly ping.

File: pandas/core/util/numba_.py (hunk 2 above)
Comment (auderson) [reply pandas/core/util/numba_.py]: A detailed description is added.

**Review by rhshadrach — APPROVED** [review]
lgtm, going to wait on merging a bit longer in case there are other reviews.

File: doc/source/whatsnew/v3.0.0.rst [hunk 3]
```diff
@@ -62,6 +62,8 @@ Other enhancements
 - Support passing a :class:`Iterable[Hashable]` input to :meth:`DataFrame.drop_duplicates` (:issue:`59237`)
 - Support reading Stata 102-format (Stata 1) dta files (:issue:`58978`)
 - Support reading Stata 110-format (Stata 7) dta files (:issue:`47176`)
+- numba apply now supports positional arguments passed as kwargs (:issue:`58995`)
```
[the remark this answers is not in the document]
Comment (auderson) [reply doc/source/whatsnew/v3.0.0.rst]: Added.

**Comment by auderson** [comment]
Also I find there are some methods like `RollingGroupby.apply` are also impacted by the changes to `RollingAndExpandingMixin.apply`. So I added more tests to cover these methods.

File: doc/source/whatsnew/v3.0.0.rst [hunk 4]
```diff
@@ -63,6 +64,7 @@ Other enhancements
 - Support reading Stata 102-format (Stata 1) dta files (:issue:`58978`)
 - Support reading Stata 110-format (Stata 7) dta files (:issue:`47176`)
 
+
```
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:67]:
```suggestion
```

File: doc/source/whatsnew/v3.0.0.rst [hunk 5]
```diff
@@ -54,6 +54,7 @@ Other enhancements
 - :meth:`Series.cummin` and :meth:`Series.cummax` now supports :class:`CategoricalDtype` (:issue:`52335`)
 - :meth:`Series.plot` now correctly handle the ``ylabel`` parameter for pie charts, allowing for explicit control over the y-axis label (:issue:`58239`)
 - :meth:`DataFrame.plot.scatter` argument ``c`` now accepts a column of strings, where rows with the same string are colored identically (:issue:`16827` and :issue:`16485`)
+- :meth:`DataFrameGroupBy.transform`, :meth:`SeriesGroupBy.transform`, :meth:`DataFrameGroupBy.agg`, :meth:`SeriesGroupBy.agg`, :meth:`RollingGroupby.apply`, :meth:`ExpandingGroupby.apply`, :meth:`Rolling.apply`, :meth:`Expanding.apply`, :meth:`DataFrame.apply`: when using numba engine in these apply methods, positional arguments now can be passed as kwargs (:issue:`58995`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:57]:
```suggestion
- :meth:`DataFrameGroupBy.transform`, :meth:`SeriesGroupBy.transform`, :meth:`DataFrameGroupBy.agg`, :meth:`SeriesGroupBy.agg`, :meth:`RollingGroupby.apply`, :meth:`ExpandingGroupby.apply`, :meth:`Rolling.apply`, :meth:`Expanding.apply`, :meth:`DataFrame.apply` with ``engine="numba"`` now supports positional arguments passed as kwargs (:issue:`58995`)
```

File: doc/source/whatsnew/v3.0.0.rst (hunk 4 above)
Comment (auderson) [reply doc/source/whatsnew/v3.0.0.rst]: This is interesting, the "+" mark was actually added by pre-commit hooks.

**Review by mroeschke — APPROVED** [review]

**Pull request merged**